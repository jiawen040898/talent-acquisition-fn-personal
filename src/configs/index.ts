export * from './auth.config';
export * from './aws.config';
export * from './skill-extract-api.config';
export * from './unleash.config';
