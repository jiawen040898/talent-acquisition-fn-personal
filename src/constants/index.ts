export * from './action-history.enum';
export * from './auth-error.enum';
export * from './cache-store.enum';
export * from './candidate.event-type.enum';
export * from './channel.enum';
export * from './contact-type.enum';
export * from './ds-lookup-api-errors.enum';
export * from './email-event.enum';
export * from './email-type.enum';
export * from './error-msg.enum';
export * from './errors';
export * from './feature-toggle-name.enum';
export * from './fit-score-recipe.enum';
export * from './identity-event.enum';
export * from './job.enum';
export * from './job-application-score-type.enum';
export * from './job-application-skill.enum';
export * from './job-application-status.enum';
export * from './person-score.enum';
export * from './person-score-event.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework.enum';
export * from './s3-errors.enum';
export * from './screening-answer-criteria.enum';
export * from './skill-extract-api-errors.enum';
export * from './talent-acquisition-event.enum';
export * from './talent-management-event.enum';
export * from './timeline.enum';
export * from './unleash-errors.enum';
export * from './user.enum';
