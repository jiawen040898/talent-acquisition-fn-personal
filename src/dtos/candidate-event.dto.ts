import { JobApplicationSkill } from '@pulsifi/interfaces';

export class CandidatePlaceResult {
    place_id?: string;
    display_address?: string;
    vicinity?: string;
    street_number?: string;
    street_name?: string;
    postal_code?: string;
    locality?: string;
    sublocality?: string;
    state?: string;
    country?: string;
    geolocation?: {
        latitude: number;
        longitude: number;
    };
}

export class CandidateProfile {
    id!: string;
    user_account_id!: number;
    first_name?: string;
    last_name?: string;
    identity_value?: string;
    identity_provider?: string | null;
    source?: string | null;
    nationality?: string | null;
    phone_number?: string | null;
    phone_code?: string | null;
    ethnicity?: string | null;
    gender?: string | 'female' | 'male' | 'others' | 'prefer_not_say' | null;
    date_of_birth?: Date | null;
    place_formatted_address?: string;
    place_result?: CandidatePlaceResult | null;
    privacy?: CandidatePrivacy | null;
    professional_summary?: string | null;
    skills?: JobApplicationSkill[] | null;
}

export class CandidatePrivacy {
    date_of_birth?: Date | null;
    ethnicity?: string | null;
    gender?: string | null;
}

export class CandidateCareer {
    id!: number;
    organization!: string;
    role!: string;
    is_current!: boolean;
    start_date?: Date;
    end_date?: Date;
    description?: string; // deprecated
    responsibilities_achievements?: string | null;
    place_formatted_address?: string | null;
    place_result?: CandidatePlaceResult | null;
}

export class CandidateEducation {
    id!: number;
    major_first?: string;
    major_second?: string;
    degree_name?: string;
    school_name?: string;
    grade_type?: string;
    grade_value?: string;
    grade_value_max?: number;
    grade_description?: string;
    grade_cgpa?: number;
    is_highest!: boolean;
    start_date?: Date;
    end_date?: Date;
    description?: string; // deprecated
    others?: string;
    achievements?: string | null;
}

export class CandidateActivity {
    id!: number;
    organization!: string;
    role!: string;
    is_role_active!: boolean;
    start_date?: Date;
    end_date?: Date;
    description?: string;
}

export class CandidateAssessmentAnswer {
    question_code!: number;
    score!: number;
    answer_at?: Date | null;
}

export class CandidateScreeningAttachmentAnswer {
    // file_type!: string;
    file_path!: string;
    file_name!: string;

    constructor(input: JsonObject) {
        // this.file_type = input['file_type'];
        this.file_path = input['file_path'];
        this.file_name = input['file_name'];
    }
}
export class CandidateScreeningQuestion {
    id!: number;
    order_no!: number;
    alias!: string;
    schema!: {
        title?: string | null;
        required?: boolean | null;
    };
}
export class CandidateScreeningAnswer {
    id!: number;
    order_no!: number;
    job_screening_question_id!: number;
    question!: CandidateScreeningQuestion;
    answer?: {
        value?:
            | string
            | string[]
            | Record<string, SafeAny>
            | CandidateScreeningAttachmentAnswer;
    };

    criteria_status!: string | 'none' | 'fail' | 'pass';
    tag?: string;
    attachment_file_path?: string;
}

export class CandidateAssessmentVideoAnswer {
    video_question_id!: number;
    question_id!: number;
    question!: string;
    video_url!: string;
    video_source_id!: number;
    thumbnail_url!: string;
}

export class CandidateAssessmentTraitResult {
    trait_id!: number;
    trait_score!: number;
    trait_alias!: string;
    trait_percentile?: number;
    trait_order?: number;
}

export class CandidateAssessmentDomainResult {
    domain_id!: number;
    domain_alias!: string;
    domain_score!: number;
    domain_percentile?: number;
    domain_order?: number;
    model_type_id?: number;
    traits?: CandidateAssessmentTraitResult[];
}

export class QuestionAnswerRaw {
    answers?: CandidateAssessmentAnswer[] | null;
    video_answers?: CandidateAssessmentVideoAnswer[] | null;
}

export class ResultRaw {
    scores?: CandidateAssessmentDomainResult[] | null;
}

export class CandidateAssessment {
    id!: number;
    questionnaire_id!: number;
    questionnaire_framework!:
        | string
        | 'personality'
        | 'work_interest'
        | 'work_value'
        | 'reasoning_logical'
        | 'reasoning_numeric'
        | 'reasoning_verbal';

    started_at?: Date;
    completed_at?: Date;
    attempts!: number;
    question_answer_raw?: QuestionAnswerRaw;
    result_raw?: ResultRaw;
}

export class CandidateApplicationCompleted {
    id!: string;
    company_id!: number;
    job_id!: string;
    job_title?: string | null;
    status?: string;
    is_questionnaires_required!: boolean;
    total_questionnaires!: number;
    total_questionnaires_completed!: number;
    is_video_required!: boolean;
    submitted_at?: Date | null;
    assessment_completed_at?: Date | null;
    completed_at?: Date;
    ext_reference_id?: string | null;
    is_anonymous!: boolean;
    profile?: CandidateProfile;
}

export class CandidateApplicationAssessmentCompleted {
    id!: string;
    company_id!: number;
    job_id!: string;
    job_title?: string | null;
    status!: string;
    is_questionnaires_required!: boolean;
    total_questionnaires!: number;
    total_questionnaires_completed!: number;
    is_video_required!: boolean;
    submitted_at?: Date | null;
    completed_at?: Date | null;
    profile!: CandidateProfile;
}

export class CandidateApplicationSubmitted extends CandidateApplicationCompleted {
    resume_original_file_path?: string | null;
    resume_file_path?: string | null;
    resume_content_path?: string | null;
    careers?: CandidateCareer[] | null;
    educations?: CandidateEducation[] | null;
    activities?: CandidateActivity[] | null;
    screening_answers?: CandidateScreeningAnswer[] | null;
    assessments?: CandidateAssessment[] | null;
}

export class CandidateApplicationAssessmentSubmitted extends CandidateAssessment {
    company_id!: number;
    job_application_id!: string;
}

export class CandidateApplicationUpdated {
    company_id!: number;
    job_id!: string;
    job_application_id?: string;
    resume_original_file_path?: string | null;
    resume_file_path?: string | null;
    resume_content_path?: string | null;
    careers?: CandidateCareer[] | null;
    screening_answers?: CandidateScreeningAnswer[] | null;
}

export class CandidateApplicationWithdrawn {
    id!: string;
    company_id!: number;
    job_id!: string;
    job_title?: string | null;
    status!: string;
    withdrawn_reason!: string;
    withdrawn_at!: Date;
}

export class CandidateApplicationAssessmentStarted extends CandidateAssessment {
    company_id!: number;
    job_application_id?: string;
}
