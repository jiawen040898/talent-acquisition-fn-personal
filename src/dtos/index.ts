export * from './audit-data.dto';
export * from './candidate-event.dto';
export * from './create-job.dto';
export * from './fit-score-recipe.dto';
export * from './geolocation.dto';
export * from './identity-update-user-event.dto';
export * from './job-application.dto';
export * from './job-application-action-history.dto';
export * from './job-application-attachment.dto';
export * from './job-application-career.dto';
export * from './job-application-contact.dto';
export * from './job-application-education.dto';
export * from './job-application-questionnaire.dto';
export * from './job-application-resume.dto';
export * from './job-application-score.dto';
export * from './job-application-screening-answer.dto';
export * from './person-score-event.dto';
export * from './place.dto';
export * from './talent-acquisition-event.dto';
export * from './talent-management-event.dto';
