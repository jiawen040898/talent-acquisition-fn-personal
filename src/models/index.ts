export * from './audit-data.entity';
export * from './entities';
export * from './job.entity';
export * from './job-application.entity';
export * from './job-application-action-history.entity';
export * from './job-application-attachment.entity';
export * from './job-application-career.entity';
export * from './job-application-contact.entity';
export * from './job-application-education.entity';
export * from './job-application-label.entity';
export * from './job-application-questionnaire.entity';
export * from './job-application-resume.entity';
export * from './job-application-score.entity';
export * from './job-application-screening-answer.entity';
export * from './job-distribution-score.entity';
export * from './job-questionnaire.entity';
export * from './job-screening-question.entity';
export * from './job-user-access.entity';
