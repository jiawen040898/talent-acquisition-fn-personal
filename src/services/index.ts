export * from './beta-distribution.service';
export * from './cognitive-score.service';
export * from './ds-lookup.service';
export * from './feature-toggle.service';
export * from './fit-score.service';
export * from './job-application.service';
export * from './job-application-score.service';
export * from './job-application-skill-extraction.service';
export * from './job-distribution-score.service';
export * from './job-user-access.service';
export * from './pairwise-score.service';
export * from './personality-score.service';
export * from './psychology.service';
export * from './publisher.service';
export * from './s3.service';
export * from './skill-extract-api.service';
