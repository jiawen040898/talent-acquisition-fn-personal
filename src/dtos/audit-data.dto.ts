import { AuditDataEntity } from '@pulsifi/models';

export class AuditDataDto {
    created_at?: Date;
    created_by?: number;
    updated_at?: Date;
    updated_by?: number;

    constructor(dto: AuditDataEntity) {
        this.created_by = dto.created_by;
        this.updated_by = dto.updated_by;
        this.created_at = dto.created_at;
        this.updated_at = dto.updated_at;
    }
}
