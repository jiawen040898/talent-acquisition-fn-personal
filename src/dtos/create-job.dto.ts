import { EmploymentType } from '@pulsifi/constants';
import { ApplicationForm } from '@pulsifi/models';

import { PlaceDto } from './place.dto';

export class CreateJobDto {
    title!: string;
    role!: string;
    employment_type!: EmploymentType;
    description?: string;
    skills?: string[];
    place_formatted_address?: string;
    place_result?: PlaceDto;
    role_fit_recipe_id?: string;
    application_form?: ApplicationForm;
    company_id?: number;
}
