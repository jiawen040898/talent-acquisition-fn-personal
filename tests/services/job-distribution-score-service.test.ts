import { ScoreData } from '@pulsifi/dtos';
import { JobDistributionScore } from '@pulsifi/models';
import { JobDistributionScoreService } from '@pulsifi/services';
import {
    reasoningLogicalScore,
    workExperienceScore,
    workStyleScore,
} from '@pulsifi/tests/fixtures/job-application-score/test-data';
import {
    reasoningLogicalDistribution,
    workStyleDistribution,
} from '@pulsifi/tests/fixtures/job-distribution-score/test-data';
import { getTestDataSource } from '@pulsifi/tests/setup';
import { DataSource, Repository } from 'typeorm';

import { setDataSource } from '../../src/database';

const spyGetLatestMeanVariance = jest.spyOn(
    JobDistributionScoreService,
    'getLatestMeanVariance',
);

describe('JobDistributionScoreService', () => {
    let dataSource: DataSource;
    let jobDistributionScoreRepo: Repository<JobDistributionScore>;

    beforeAll(async () => {
        dataSource = await getTestDataSource();
        jobDistributionScoreRepo =
            dataSource.getRepository(JobDistributionScore);

        setDataSource(dataSource);

        await jobDistributionScoreRepo.save([
            workStyleDistribution,
            reasoningLogicalDistribution,
        ]);
    });

    describe('updateJobDistributionScore', () => {
        describe('when score type is work style', () => {
            beforeAll(async () => {
                // Arrange
                jest.clearAllMocks();

                // Act
                await JobDistributionScoreService.updateJobDistributionScore(
                    workStyleDistribution.job_id,
                    workStyleScore as ScoreData,
                );
            });

            it('should call getLatestMeanVariance with correct number', () => {
                // Assert
                expect(spyGetLatestMeanVariance).toHaveBeenCalledWith(
                    expect.objectContaining({
                        id: workStyleDistribution.id,
                    }),
                    expect.closeTo(0.6010443),
                );
            });

            it('should update job distribution score', async () => {
                // Assert
                const updatedDistributionScore =
                    await jobDistributionScoreRepo.findOneOrFail({
                        select: ['size', 'mean', 'variance', 'alpha', 'beta'],
                        where: {
                            job_id: workStyleDistribution.job_id,
                            score_type: workStyleDistribution.score_type,
                        },
                    });
                expect(updatedDistributionScore.size).toEqual(
                    workStyleDistribution.size + 1,
                );
                expect(updatedDistributionScore.mean).not.toEqual(
                    workStyleDistribution.mean,
                );
                expect(updatedDistributionScore).toMatchSnapshot();
            });
        });

        describe('when score type is reasoning logical', () => {
            beforeAll(async () => {
                // Arrange
                jest.clearAllMocks();

                // Act
                await JobDistributionScoreService.updateJobDistributionScore(
                    reasoningLogicalDistribution.job_id,
                    reasoningLogicalScore as ScoreData,
                );
            });

            it('should call getLatestMeanVariance with correct number', () => {
                // Assert
                expect(spyGetLatestMeanVariance).toHaveBeenCalledWith(
                    expect.objectContaining({
                        id: reasoningLogicalDistribution.id,
                    }),
                    expect.closeTo(0.77832782),
                );
            });

            it('should update job distribution score', async () => {
                // Assert
                const updatedDistributionScore =
                    await jobDistributionScoreRepo.findOneOrFail({
                        select: ['size', 'mean', 'variance', 'alpha', 'beta'],
                        where: {
                            job_id: reasoningLogicalDistribution.job_id,
                            score_type: reasoningLogicalDistribution.score_type,
                        },
                    });
                expect(updatedDistributionScore.size).toEqual(
                    reasoningLogicalDistribution.size + 1,
                );
                expect(updatedDistributionScore.mean).not.toEqual(
                    reasoningLogicalDistribution.mean,
                );
                expect(updatedDistributionScore).toMatchSnapshot();
            });
        });

        describe('when score type is work experience', () => {
            beforeAll(async () => {
                // Arrange
                jest.clearAllMocks();

                // Act
                await JobDistributionScoreService.updateJobDistributionScore(
                    reasoningLogicalDistribution.job_id,
                    workExperienceScore as ScoreData,
                );
            });

            it('should not call getLatestMeanVariance', () => {
                // Assert
                expect(spyGetLatestMeanVariance).not.toHaveBeenCalled();
            });
        });
    });

    describe('getLatestMeanVariance', () => {
        it('should return correct output', () => {
            // Act
            const result = JobDistributionScoreService.getLatestMeanVariance(
                workStyleDistribution,
                0.78,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should still calculate if score is higher than 1', () => {
            // Act
            const result = JobDistributionScoreService.getLatestMeanVariance(
                workStyleDistribution,
                1.1,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should still calculate if score is lower than 0', () => {
            // Act
            const result = JobDistributionScoreService.getLatestMeanVariance(
                workStyleDistribution,
                -0.1,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });
});
