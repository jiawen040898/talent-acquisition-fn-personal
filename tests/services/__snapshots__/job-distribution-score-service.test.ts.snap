// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobDistributionScoreService getLatestMeanVariance should return correct output 1`] = `
{
  "alpha": 17.587782413153192,
  "beta": 5.457045630861126,
  "mean": 0.7631986829999999,
  "variance": 0.007516229807767551,
}
`;

exports[`JobDistributionScoreService getLatestMeanVariance should still calculate if score is higher than 1 1`] = `
{
  "alpha": 12.464704142046196,
  "beta": 3.698062232966139,
  "mean": 0.7711986829999999,
  "variance": 0.010281050879767554,
}
`;

exports[`JobDistributionScoreService getLatestMeanVariance should still calculate if score is lower than 0 1`] = `
{
  "alpha": 4.801203987402785,
  "beta": 1.676416787596333,
  "mean": 0.7411986829999999,
  "variance": 0.025652971859767545,
}
`;

exports[`JobDistributionScoreService updateJobDistributionScore when score type is reasoning logical should update job distribution score 1`] = `
JobDistributionScore {
  "alpha": 17.592317293005138,
  "beta": 5.459715373703165,
  "mean": 0.7631568784999999,
  "size": 1873,
  "variance": 0.007514893223487087,
}
`;

exports[`JobDistributionScoreService updateJobDistributionScore when score type is work style should update job distribution score 1`] = `
JobDistributionScore {
  "alpha": 16.290701014247517,
  "beta": 5.180458513190539,
  "mean": 0.7587247905,
  "size": 1873,
  "variance": 0.0081465080854954,
}
`;
